formats/json/commonMain/src/kotlinx/serialization/json/JsonNamingStrategy.kt

```diff
@@ -0,0 +1,120 @@
+package kotlinx.serialization.json
+
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+
+
+/**
+ * Represents naming strategy — a transformer for serial names in a [Json] format.
+ * Transformed serial names are used for both serialization and deserialization.
+ * Actual transformation happens in the [serialNameForJson] function.
+ * A naming strategy is always applied globally in the Json configuration builder
+ * (see [JsonBuilder.namingStrategyForEnums] and [JsonBuilder.namingStrategyForProperties]).
+ * However, it is possible to apply additional filtering inside the transformer using the `descriptor` parameter in [serialNameForJson].
+ *
+ * Original serial names are never used after transformation, so they are ignored in a Json input.
+ * If the original serial name is present in the Json input but transformed is not,
+ * [MissingFieldException] still would be thrown. If one wants to preserve the original serial name for deserialization,
+ * one should use the [JsonNames] annotation.
+ *
+ * ### Common pitfalls in conjunction with other Json features
+ *
+ * * Due to the nature of kotlinx.serialization framework, naming strategy transformation is applied to all properties regardless
+ * of whether their serial name was taken from the property name or provided by @[SerialName] annotation.
```

Comment:
This is probably very subjective, but it seems very unobvious to me that the value from the `SerialName` is also transformed.
In addition, the user does not have any opportunity to fix the field name rigidly (for example, some kind of naming exception due to legacy).

If  we leave it like this, I think it would be good to correct the KDoc for the `SerialName` so that it would not be a surprise to the user that the annotation does not always specify the final name of the field.